### ISSUE TITLE
unison wire protocol depends on ocaml version

### ISSUE THREAD

**Issue by gdt** [bcpierce00/unison#375]
unison instead should have a defined protocol that is invariant across ocaml versions

(I suspect everyone agrees, but that it's simply hard.  Speak up if you see it differently.)

The point of this issue is to reference when closing issues that are specific instances of this problem.

(Please do not add comments that you have this problem too, or that some packaging system has an old version of ocaml. The problem is well understood and packaging system choices are out of scope for the unison issue tracker.)

**Comment by bcpierce00** [comment]
Stéphane actually says he has an OCaml-version-agnostic wire format coming soon!

**Comment by glondu** [comment]
> Stéphane actually says he has an OCaml-version-agnostic wire format coming soon!

Indeed. So far, I've removed the dependency to Marshal, it's available in this branch: https://github.com/glondu/unison/tree/umarshal . There remains the dependency on Digest.

I'm waiting for the (hopefully imminent) release (2.51.3 or 2.52.0 or whatever) to be done before making pull requests.

**Comment by gdt** [comment]
I saw that - great news.  I just needed an issue to point to when closing other issues that are basically dups of the big picture.

**Comment by breakone9r** [comment]
I have what appears, on the surface, to be this problem. However, I have ensured that unison is the same version, and compiled from the same version of ocaml. 4.05.0 in this case (its the latest version I can get on my freebsd-13 box, as later versions will not compile with clang 11.0 without patches.) and yet I still get:
```break19@notaryzen ~ % unison mudlet -version
unison version 2.51.2 (ocaml 4.05.0)
break19@notaryzen ~ % unison mudlet         
Unison 2.51.2 (ocaml 4.05.0): Contacting server...
Connected [//notaryzen//home/break19/.config/mudlet -> //zenstation//usr/home/break19/.config/mudlet]
Looking for changes
Warning: No archive files were found for these roots, whose canonical names are:
        /home/break19/.config/mudlet
        //zenstation//usr/home/break19/.config/mudlet
This can happen either
because this is the first time you have synchronized these roots, 
or because you have upgraded Unison to a new version with a different
archive format.  

Update detection may take a while on this run if the replicas are 
large.

Unison will assume that the 'last synchronized state' of both replicas
was completely empty.  This means that any files that are different
will be reported as conflicts, and any files that exist only on one
replica will be judged as new and propagated to the other replica.
If the two replicas are identical, then no changes will be reported.

If you see this message repeatedly, it may be because one of your machines
is getting its address from DHCP, which is causing its host name to change
between synchronizations.  See the documentation for the UNISONLOCALHOSTNAME
environment variable for advice on how to correct this.

Donations to the Unison project are gratefully accepted: 
http://www.cis.upenn.edu/~bcpierce/unison

Press return to continue.[<spc>] Uncaught exception Failure("input_value: ill-formed message")
Raised at file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/lwt/lwt.ml", line 126, characters 16-23
Called from file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/lwt/generic/lwt_unix_impl.ml", line 102, characters 8-23
Called from file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/update.ml" (inlined), line 2105, characters 2-69
Called from file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/uitext.ml", line 720, characters 16-56
Called from file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/uitext.ml", line 808, characters 6-90
Called from file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/uitext.ml", line 830, characters 19-66
Called from file "/home/break19/.opam/4.05.0/.opam-switch/build/unison.2.51.2/src/uitext.ml", line 886, characters 21-43

break19@notaryzen ~ % ssh zenstation 
Last login: Sat Oct 10 17:56:36 2020 from 192.168.0.130
FreeBSD 13.0-CURRENT (GENERIC-NODEBUG) #1 494ba3d79-c119(master): Fri Oct  9 16:11:27 CDT 2020

Welcome to FreeBSD!

Release Notes, Errata: https://www.FreeBSD.org/releases/
Security Advisories:   https://www.FreeBSD.org/security/
FreeBSD Handbook:      https://www.FreeBSD.org/handbook/
FreeBSD FAQ:           https://www.FreeBSD.org/faq/
Questions List: https://lists.FreeBSD.org/mailman/listinfo/freebsd-questions/
FreeBSD Forums:        https://forums.FreeBSD.org/

Documents installed with the system are in the /usr/local/share/doc/freebsd/
directory, or can be installed later with:  pkg install en-freebsd-doc
For other languages, replace "en" with a language code like de or fr.

Show the version of FreeBSD installed:  freebsd-version ; uname -a
Please include that output and any error messages when posting questions.
Introduction to manual pages:  man man
FreeBSD directory layout:      man hier

Edit /etc/motd.template to change this login announcement.
box1 ~ % /usr/home/break19/.opam/default/bin/unison -version
unison version 2.51.2 (ocaml 4.05.0)```

```box2 ~ % /home/break19/.opam/4.05.0/bin/unison -version
unison version 2.51.2 (ocaml 4.05.0)```

**Comment by gdt** [comment]
When we have a new protocol, it should have a version number, even if we don't think we will have to change it, to enable compat or at least detection.

**Comment by glondu** [comment]
@breakone9r Are you sure the right unison binary is invoked on the server side? You can select it with the servercmd preference.

**Comment by breakone9r** [comment]
> @breakone9r Are you sure the right unison binary is invoked on the server side? You can select it with the servercmd preference.

The only version installed on the FreeBSD box is the version listed above.

I'm running unison from the opensuse box, and invoking it via the correct path.. so I'm fairly sure...

I've also tried using pure sockets, by running it as a daemon on the FreeBSD side... Same error

**Comment by gdt** [comment]
I suspect then that you have run into yet another issue, which is that the wire format can change based on something else we don't know about.

BTW you can probably grab the pkgsrc entry for ocaml, which I am pretty sure builds with clang, if you want to hack on ocaml.

**Comment by breakone9r** [comment]
> I suspect then that you have run into yet another issue, which is that the wire format can change based on something else we don't know about.
> 
> BTW you can probably grab the pkgsrc entry for ocaml, which I am pretty sure builds with clang, if you want to hack on ocaml.

It builds fine with clang up to version 11. The default compiler on FreeBSD 13 is 11. :/

I've been playing around in the ports though, there's a couple of patches for 11 though.

I've also pinged the official FreeBSD maintainer, and haven't gotten a response yet. I'm just a poor luser though. Lol. I've tried a few times building it outside of opam. I may give it another good go this weekend though.

**Comment by mikeperalta1** [comment]
> > Stéphane actually says he has an OCaml-version-agnostic wire format coming soon!
> 
> Indeed. So far, I've removed the dependency to Marshal, it's available in this branch: https://github.com/glondu/unison/tree/umarshal . There remains the dependency on Digest.
> 
> I'm waiting for the (hopefully imminent) release (2.51.3 or 2.52.0 or whatever) to be done before making pull requests.

Hello and thank you for your efforts. I was unable to get the "umarshal" branch to compile on Centos 8 (Linux [hostname] 4.18.0-305.7.1.el8_4.x86_64 #1 SMP Tue Jun 29 21:55:12 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux)

```
make -C src UISTYLE=text
make[1]: Entering directory '/usr/local/share/repos/unison-glondu/src'
UISTYLE = text
Building for Unix
NATIVE = true
THREADS = false
STATIC = false
OSTYPE =
OSARCH = Linux
ocamlopt: ubase/umarshal.ml ---> ubase/umarshal.cmx
ocamlopt -g -I lwt -I ubase -I system -I fsmonitor -I fsmonitor/linux -I fsmonitor/windows -I system/generic -I lwt/generic -c /usr/local/share/repos/unison-glondu/src/ubase/umarshal.ml
File "/usr/local/share/repos/unison-glondu/src/ubase/umarshal.ml", line 34, characters 12-30:
Error: Unbound value Bytes.get_int64_be
make[1]: *** [Makefile.OCaml:423: ubase/umarshal.cmx] Error 2
make[1]: Leaving directory '/usr/local/share/repos/unison-glondu/src'
make: *** [Makefile:6: text] Error 2
```
Local (Ubuntu) ocaml is 4.08.1-10, remote (CentOS) ocal is 4.07.0-3.el8

Sorry for postig the report here. Your fork didn't have issues enabled.

I'm going to try and compile ocaml from source, but for now unison is broken for me :'(

**Comment by yawaramin** [comment]
@mikeperalta1 as per https://ocaml.org/api/Bytes.html#VALget_int64_be , `Bytes.get_int64_be` is available since OCaml version 4.08.

You don't need to compile OCaml from source on CentOS. Just install [opam](http://opam.ocaml.org/doc/Install.html#Fedora-CentOS-and-RHEL) (`dnf install opam`), set it up (`opam init` etc.), and you should have either a recent OCaml switch (version) or be able to create one easily.

Also, it looks like you are using the `Makefile`? Did `dune` not work?

**Comment by mikeperalta1** [comment]
@yawaramin Thank you but I'm okay now. Building ocaml from source allowed me to compile unison correctly.

As for ```dune```, I'm a huge newb and don't even know what that is.

**Comment by yawaramin** [comment]
@mikeperalta1 glad it works now. Dune is the standard OCaml build system: https://dune.build/

**Comment by isoboroff** [comment]
Also seeing this, syncing between a macOS and a Linux machine, so assuming that everything is in harmony is just not realistic.
```
  Waiting for changes from server                                       
Reconciling changes
Fatal error: Server: Fatal error during unmarshaling (input_value: ill-formed message),
possibly because client and server have been compiled with different versions of the OCaml compiler.
```

**Comment by nekohayo** [comment]
For people interested in keeping an eye on progress towards an OCaml-version-independent Unison (which seems to be what this issue—and the other two pinned issues—is about), is there a particular branch or repository where the work is occurring nowadays? @glondu's branch at https://github.com/glondu/unison/commits/umarshal appears unchanged since 2020; above it was mentioned that it was blocked on unison making new releases, but that has happened since then. Or was that branch merged already?

In other words, what would be a good indicator of the current status of work towards a version of Unison that is not as OCaml-version-dependent as it has historically been?

**Comment by gdt** [comment]
Arguably the right place to listen is the hackers@ list.    The umarshal branch has not been merged, but there has also been no decision not to merge it.
